include/ginkgo/core/solver/bicgstab.hpp

```diff
@@ -134,7 +141,9 @@ class Bicgstab : public EnableLinOp<Bicgstab<ValueType>>,
           parameters_{factory->get_parameters()},
           system_matrix_{std::move(system_matrix)}
     {
-        if (parameters_.preconditioner) {
+        if (parameters_.generated_preconditioner) {
```

Comment:
I am not sure, but I think in some cases you would like to regenerate the preconditioner and pass in a `LinOpFactory`. I am not sure if this strict dependency is suitable.

I would prefer a parameter, something like `always_use_generated_preconditioner` or something like that. I know that this increases the list of parameters, but I am not sure if there is another way.

Reply:
If you want the preconditioner to be generated, you could simply not use the `with_generated_preconditioner`. If you provide both, one has to be chosen above another. Adding another factory parameter would not really make the situation better since you could just choose to give one or the other instead.

We could resolve this by providing a `set_preconditioner()` function for all solvers, however, that can lead undefined behavior when a preconditioner is set while an `apply` is running (also, one preconditioner is generated in the beginning without being used).
Despite that, it would allow you to use the same factory for a lot of solvers, while allowing you to change the preconditioner. The downsides are still totally manageable (it should be clear that changing the preconditioner while calculating is a bad idea).

Reply:
I think the way it has been implemented currently is good enough, though a `set_preconditioner` method might be useful as well. I dont think there is a problem of setting the preconditioner when the apply is running, because each apply call is synchronous (?), atleast currently.

Reply:
Yes, but the problem is that we might want to go asynchronous at some point, and we can't simply take this functionality back.

Reply:
Just for reference: After the discussion, we agreed that I should add both the factory parameter (which is already implemented) and the `set_preconditioner` function.

Reply:
I guess not every solver has the "set_preconditioner" function - i.e. only "preconditionable" objects.

Reply:
Yes, that is how I plan to implement it.
But I will also add a `set_solver` for Iterative Refinement since the interface there is very much like the ones for preconditioner.